rir/src/ir/Compiler.cpp

```diff
@@ -35,33 +36,89 @@ class Context {
       public:
         CodeStream cs;
         std::stack<LoopContext> loops;
-        CodeContext(SEXP ast, FunctionHandle& fun) : cs(fun, ast) {}
+        CodeContext* parent;
+        CodeContext(SEXP ast, FunctionHandle& fun, CodeContext* p) : cs(fun, ast), parent(p) {}
+        virtual ~CodeContext() {}
+        bool inLoop() {
+            return !loops.empty() ||
+                    (parent && parent->inLoop());
+        }
+        LabelT loopNext() {
+            return !loops.empty() ? loops.top().next_ : parent->loopNext();
+        }
+        LabelT loopBreak() {
+            return !loops.empty() ? loops.top().break_ : parent->loopBreak();
+        }
+        void setContextNeeded() {
+            if (loops.empty() && parent)
+                parent->setContextNeeded();
+            else
+                loops.top().context_needed_ = true;
+        }
+        virtual bool gotoOK() {
```

Comment:
can you give this a more meaningful name (maybe loopIsLocal() or knownLoopHeaderPos)